Add support for trusting additional certificates at runtime

Hakky54 opened the pull request.

henryju commented on sslcontext-kickstart/src/main/java/nl/altindag/ssl/trustmanager/CompositeX509ExtendedTrustManager.java:
```diff
@@ -98,7 +93,10 @@ public void checkServerTrusted(X509Certificate[] chain, String authType, SSLEngi
     @Override
     public X509Certificate[] getAcceptedIssuers() {
-        return Arrays.copyOf(acceptedIssuers, acceptedIssuers.length);
+        return trustManagers.stream()
```
:+1:

henryju commented on sslcontext-kickstart/src/main/java/nl/altindag/ssl/trustmanager/InflatableX509ExtendedTrustManager.java:
```diff
@@ -39,20 +52,95 @@ public class InflatableX509ExtendedTrustManager extends HotSwappableX509Extended
+        } catch (CertificateException e) {
+            boolean shouldBeTrusted = certificateAndAuthTypeTrustPredicate.test(chain, authType);
+            if (shouldBeTrusted) {
+                addCertificates(Collections.singletonList(chain[0]));
```
And finally the `addCertificates` will probably be in a dead lock state, since I think the current thread can be upgraded from read to write lock, but since there are many threads owning the read lock, I don't think it should work.

Hakky54 replied:
I am not quite sure about that. I think it is still ok to have a readlock within the scope of this method as it is no issue to for all threads to call `trustManagerConsumer.checkTrusted(this);` However when there is a Certificate exception and it enters the catch block other threads needs to wait and that is now happening thanks to the synchronized block. Within that block it will enter the addCertificates method which contains a write lock. That method can also be accessed outside of this class so anyone needs to wait till that execution is finished if it is blocked. So my feeling is that there is no dead lock, but I would like to get your insight after testing it with the latest changes

henryju reviewed (commented):
The API looks exactly like what we would need. I only have doubt regarding the thread safety for the predicate. IMO it should be guaranteed that the predicate is not called concurrently, and that in the case where a first request leads to the certificate being added to the store, other concurrent requests should also trust this certificate without calling the predicate again.
Maybe I missed something, I will try to do a real test.

Hakky54 commented on sslcontext-kickstart/src/main/java/nl/altindag/ssl/trustmanager/InflatableX509ExtendedTrustManager.java:
```diff
@@ -39,20 +52,95 @@ public class InflatableX509ExtendedTrustManager extends HotSwappableX509Extended
+    }
+
+    private void checkTrusted(TrustManagerConsumer trustManagerConsumer, X509Certificate[] chain, String authType) throws CertificateException {
+        readLock.lock();
```
Why wouldn't the lock prevent the concurrent thread? To be honest, I am not an expect on the concurrency topic. So feel free to provide your input regarding the overall thread safety solution for the changes in this PR. Let's make it better together

Hakky54 commented on sslcontext-kickstart/src/main/java/nl/altindag/ssl/trustmanager/InflatableX509ExtendedTrustManager.java:
```diff
@@ -39,20 +52,95 @@ public class InflatableX509ExtendedTrustManager extends HotSwappableX509Extended
+        try {
+            trustManagerConsumer.checkTrusted(this);
+        } catch (CertificateException e) {
+            boolean shouldBeTrusted = certificateAndAuthTypeTrustPredicate.test(chain, authType);
```
As far as I know the readlock can be acquired by multiple threads as long as they don't change the data. The writer lock can be acquired by a single thread and others need to wait till it is finished. So coming back to your concern, I think the predicate can be called by multiple different threads, but that is something which we should prevent, right?

You want to provide the user a single pop-up with asking to trust the certificate at a time, right?

henryju replied:
In case you want to avoid the synchronize block, I made an attempt to use double checked locking: https://github.com/Hakky54/sslcontext-kickstart/pull/342/commits/5d1556d2d1a07e0ed534d3611cc37f4299f56a94

Also I confirm that trying to upgrade a read lock to a write lock will cause dead locks, so a read lock should always be released before trying to take a write lock.

Hakky54 commented on sslcontext-kickstart/src/main/java/nl/altindag/ssl/trustmanager/InflatableX509ExtendedTrustManager.java:
```diff
@@ -129,13 +129,14 @@ private void checkTrusted(TrustManagerConsumer trustManagerConsumer, X509Certifi
-            if (shouldBeTrusted) {
-                addCertificates(Collections.singletonList(chain[0]));
-                return;
+            synchronized (this) {
```
Ah, yes I see it what you are trying to do and it is getting better, however multiple threads are able to call checkTrusted while they need to wait before entering the exception block. What will happen when user trust the untrested certificate but another thread is having the same certificate. The user already accepted the certificate, however the other thread was waiting till the catch block was accesible again and will prompt the user with the same question for the same certificate.

What do you think of addind the syncronize block to the method. It will therefor allow only 1 method at a time to enter and do the certificate checking and if is is untrusted the user will be prompt and if trusted, the another thread with the same certificate will pass instantly as it will call the checktrusted after the trustmanager is getting updated. Maybe something like this:

```java
private synchronized void checkTrusted(TrustManagerConsumer trustManagerConsumer, X509Certificate[] chain, String authType) throws CertificateException {
    try {
        trustManagerConsumer.checkTrusted(this);
    } catch (CertificateException e) {
        boolean shouldBeTrusted = certificateAndAuthTypeTrustPredicate.test(chain, authType);
        if (shouldBeTrusted) {
            addCertificates(Collections.singletonList(chain[0]));
        } else {
            throw e;
        }
    }
}
```

What do you think, does it make sense?

henryju replied:
> What do you think of addind the syncronize block to the method
> What do you think, does it make sense?

If think this is fine, and pretty safe, since this is more or less what is done on IntelliJ [side](https://github.com/JetBrains/intellij-community/blob/f3ac42aab02f84a92afbfc1fb48ae15906e07297/platform/platform-api/src/com/intellij/util/net/ssl/ConfirmingTrustManager.java#L216). It might be a performance issue in situation where there are many concurrent requests, that's why I suggested the [double-checked locking](https://en.wikipedia.org/wiki/Double-checked_locking). But for us, both will be fine, since we don't expect a high frequency of requests.

Hakky54 replied:
I didn't know about the double-checked locking. I am able to understand what you did and I think it is a nice solution. My own code snippet causes more overhead and is less performing and can have performance issues for some end-user.

I would like to merge your PR into this branch, however it has some merge conflicts. Can you resolve those on your side and push your commit?

henryju replied:
Done

Hakky54 commented:
I added some tests and also adjusted the TrustManager. Can you maybe have another look at it? Strangly I cannot get the 100% coverage, although during debugging I can see it is covering that line but sonarcloud is reporting as not covered, see here: https://sonarcloud.io/code?id=io.github.hakky54%3Asslcontext-kickstart-parent&pullRequest=338&selected=io.github.hakky54%3Asslcontext-kickstart-parent%3Asslcontext-kickstart%2Fsrc%2Fmain%2Fjava%2Fnl%2Faltindag%2Fssl%2Ftrustmanager%2FInflatableX509ExtendedTrustManager.java&line=138

Maybe you know what I am missing here.

By the way, did you got any chance to try out this branch within sonarlint-core and if so do you need anything else to be added?

henryju commented:
Hi @Hakky54 

I made a final test and there is only one minor fix that would be useful to us: when the keystone file path is in a subfolder that doesn't exists yet, then the InflatableTrustManager fails.
We could create the folder in advance, but this is better to only create it when a custom certificate needs to be trusted for the first time.

Something like:

```
diff --git a/sslcontext-kickstart/src/main/java/nl/altindag/ssl/util/internal/IOUtils.java b/sslcontext-kickstart/src/main/java/nl/altindag/ssl/util/internal/IOUtils.java
index f9d802c..674d043 100644
--- a/sslcontext-kickstart/src/main/java/nl/altindag/ssl/util/internal/IOUtils.java
+++ b/sslcontext-kickstart/src/main/java/nl/altindag/ssl/util/internal/IOUtils.java
@@ -96,6 +96,7 @@ public final class IOUtils {
     }
 
     public static void write(Path path, Consumer<OutputStream> consumer) {
+        mkDirs(path.getParent());
         try (OutputStream outputStream = Files.newOutputStream(path, StandardOpenOption.CREATE)) {
             consumer.accept(outputStream);
         } catch (Exception e) {
@@ -103,6 +104,14 @@ public final class IOUtils {
         }
     }
 
+    private static void mkDirs(Path path) {
+        try {
+            Files.createDirectories(path);
+        } catch (IOException e) {
+            throw new GenericIOException(e);
+        }
+    }
+
     @FunctionalInterface
     public interface Consumer<T> {
         void accept(T t) throws Exception;
```

Except from that, all out test suite is green.

Hakky54 commented:
Excellent, I will apply a fix for creating a subdirectory if it does not exists yet and make a release this evening. I will keep you updated.

Hakky54 commented:
These changes are now available in version 8.1.0
Thank you mate for working together on this PR
